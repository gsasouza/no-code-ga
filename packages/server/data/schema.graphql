"""Algorithm data"""
type Algorithm implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  user: User
  setup: Setup
}

"""A connection to a list of items."""
type AlgorithmConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [AlgorithmEdge]
  count: Int
}

"""An edge in a connection."""
type AlgorithmEdge {
  """The item at the end of the edge"""
  node: Algorithm

  """A cursor for use in pagination"""
  cursor: String!
}

input AuthUserInput {
  token: String!
  clientMutationId: String
}

type AuthUserPayload {
  me: User
  token: String
  error: String
  clientMutationId: String
}

type DataModel {
  name: String
  type: String
}

type Mutation {
  UserAuth(input: AuthUserInput!): AuthUserPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node
  me: User
  users(after: String, first: Int, before: String, last: Int, search: String): UserConnection
  algorithms(after: String, first: Int, before: String, last: Int, search: String): AlgorithmConnection
}

"""Setup data"""
type Setup {
  populationSize: Int
  generateFunction: String
  testFunction: String
  dataModel: [DataModel]
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  email: String
  profilePicture: String
}

"""A connection to a list of items."""
type UserConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [UserEdge]
  count: Int
}

"""An edge in a connection."""
type UserEdge {
  """The item at the end of the edge"""
  node: User

  """A cursor for use in pagination"""
  cursor: String!
}
