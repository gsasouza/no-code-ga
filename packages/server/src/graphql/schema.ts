import { GraphQLSchema } from 'graphql';

import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

export default schema;
