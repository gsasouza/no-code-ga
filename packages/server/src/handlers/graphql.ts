import { ApolloServer } from 'apollo-server-lambda';
import { applyMiddleware } from 'graphql-middleware';

import { MONGO_URL, getConnection } from '../common';
import { dataloadersMiddleware } from '../graphql/loaders';
import schema from '../graphql/schema';

const server = new ApolloServer({
  schema: applyMiddleware(schema, dataloadersMiddleware),
  context: async () => {
    const dbConnection = await getConnection(MONGO_URL);
    return { dbConnection };
  },
  playground: {
    endpoint: '/dev/graphql',
  },
  introspection: true,
});

exports.handler = server.createHandler({
  cors: {
    origin: '*',
    credentials: true,
  },
});
