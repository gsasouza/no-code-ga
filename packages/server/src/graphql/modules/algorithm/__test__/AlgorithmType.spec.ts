import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createAlgorithm,
  getContext,
  createUser,
} from '../../../../../test/helpers';
import schema from '../../../schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('Algorithm Type', () => {
  it('should get an algorithm', async () => {
    const user = await createUser();
    const algorithm = await createAlgorithm({ user });

    // language=GraphQL
    const query = `
      query Q($id: ID!) {
        algorithm: node(id: $id) {
          id
          ... on Algorithm {
            name
            user {
              id
            }
            setup {
              dataModel {
                name
                type
              }
              generateFunction
              testFunction
              populationSize
            }
          }
        }
      }
    `;

    const variables = {
      id: toGlobalId('Algorithm', algorithm._id),
    };

    const rootValue = {};
    const context = await getContext({ user });
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.data).toMatchSnapshot({
      algorithm: {
        id: expect.any(String),
        user: {
          id: expect.any(String),
        },
      },
    });
  });

  it('should get a list of algorithms', async () => {
    const user = await createUser();

    await createAlgorithm({ user });
    await createAlgorithm({ user });
    await createAlgorithm({ user });
    await createAlgorithm({ user });
    // language=GraphQL
    const query = `
      query Q {
        algorithms {
          edges {
            node {
              name
              user {
                id
              }
              setup {
                dataModel {
                  type
                  name
                }
                generateFunction
                testFunction
                populationSize
              }
            }
          }
        }
      }
    `;

    const variables = {};

    const rootValue = {};
    const context = await getContext({ user });
    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.data).toMatchSnapshot();
  });
});
