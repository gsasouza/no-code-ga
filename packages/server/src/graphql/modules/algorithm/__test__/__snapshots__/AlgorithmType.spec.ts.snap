// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Algorithm Type should get a list of algorithms 1`] = `
Object {
  "algorithms": Object {
    "edges": Array [
      Object {
        "node": Object {
          "name": "Algorithm#3",
          "setup": Object {
            "dataModel": Array [
              Object {
                "name": "prop",
                "type": "Number",
              },
            ],
            "generateFunction": "(x) => x",
            "populationSize": 10,
            "testFunction": "(x) => x",
          },
          "user": Object {
            "id": "VXNlcjo=",
          },
        },
      },
      Object {
        "node": Object {
          "name": "Algorithm#2",
          "setup": Object {
            "dataModel": Array [
              Object {
                "name": "prop",
                "type": "Number",
              },
            ],
            "generateFunction": "(x) => x",
            "populationSize": 10,
            "testFunction": "(x) => x",
          },
          "user": Object {
            "id": "VXNlcjo=",
          },
        },
      },
      Object {
        "node": Object {
          "name": "Algorithm#1",
          "setup": Object {
            "dataModel": Array [
              Object {
                "name": "prop",
                "type": "Number",
              },
            ],
            "generateFunction": "(x) => x",
            "populationSize": 10,
            "testFunction": "(x) => x",
          },
          "user": Object {
            "id": "VXNlcjo=",
          },
        },
      },
      Object {
        "node": Object {
          "name": "Algorithm#1",
          "setup": Object {
            "dataModel": Array [
              Object {
                "name": "prop",
                "type": "Number",
              },
            ],
            "generateFunction": "(x) => x",
            "populationSize": 10,
            "testFunction": "(x) => x",
          },
          "user": Object {
            "id": "VXNlcjo=",
          },
        },
      },
    ],
  },
}
`;

exports[`Algorithm Type should get an algorithm 1`] = `
Object {
  "algorithm": Object {
    "id": Any<String>,
    "name": "Algorithm#1",
    "setup": Object {
      "dataModel": Array [
        Object {
          "name": "prop",
          "type": "Number",
        },
      ],
      "generateFunction": "(x) => x",
      "populationSize": 10,
      "testFunction": "(x) => x",
    },
    "user": Object {
      "id": Any<String>,
    },
  },
}
`;
